[fix] 새로고침 시 401 오류 해결

kimhm0828 opened the pull request:
## ✅ 연관된 이슈
#52 

## 📝 작업 내용
- axios 요청 인터셉터에서 토큰 만료 여부 확인 후, 토큰이 만료되거나 없으면 refresh 요청을 보내도록 수정
- 응답 인터셉터의 refresh 로직은 필요없지만 혹시모를 예외를 위해 남겨둠 

## 🖼️ 스크린샷 (선택)

kimhm0828 committed:
[feat] 관리자 페이지 구현

kimhm0828 committed:
Merge branch 'develop' of https://github.com/woori-fisa-final-project/woori-learn-front into feat/#38

kimhm0828 committed:
[fix] 새로고침 시 401 오류 해결

kimhm0828 committed:
[chore] 필요없는 파일 삭제

kimhm0828 committed:
Merge branch 'develop' of https://github.com/woori-fisa-final-project/woori-learn-front into fix/#52

kimhm0828 committed:
[fix] 경쟁 상태 해결 및 디코딩 오류 발생 가능성 방지

kimhm0828 commented:
/gemini review

kimingzero commented on src/utils/axiosInstance.ts:
```diff
@@ -23,12 +24,45 @@ const axiosInstance = axios.create({
+            window.location.href = "/login";
+            throw new ApiError(401, "토큰 갱신 실패");
+          } finally {
+            isRefreshing = false;
```
1. refreshPromise 종료 시점이 불안정함 — race condition 위험
문제
```
finally {
  isRefreshing = false;
}
```

request 인터셉터에서 refreshPromise가 끝나면 바로 isRefreshing을 false로 돌림.
하지만 response 인터셉터에서도 isRefreshing을 조작함.

➡ 한 곳에서만 관리되어야 하는 shared state를 여러 로직에서 만지고 있음 → race condition 위험

해결 가이드

isRefreshing / refreshPromise 상태는 "response 인터셉터"에서만 관리하도록 일원화시키는 것이 좋음.

request 인터셉터에서는 단순히 refreshPromise만 확인하도록 리팩터링 필요.

kimingzero commented on src/utils/axiosInstance.ts:
```diff
@@ -23,12 +24,45 @@ const axiosInstance = axios.create({
     }
-    const token = useAuthStore.getState().accessToken;
+
+    let token = useAuthStore.getState().accessToken;
```
토큰이 null일 때도 isTokenExpired(token)을 호출하는 구조 위험
```
let token = useAuthStore.getState().accessToken;

if(isTokenExpired(token)){
```

토큰이 null이면 isTokenExpired 내부에서 예외 발생할 가능성 높음.

개선
```
if (!token) {
  return config;   // 혹은 로그인 화면으로 redirect
}

if (isTokenExpired(token)) {
  ...
}
```

kimingzero commented on src/utils/axiosInstance.ts:
4. originalRequest.headers null 체크 필요

axios config는 Node/Browser 환경마다 header가 undefined일 수 있음.
```
originalRequest.headers["Authorization"] = `Bearer ${newAccessToken}`;
```

이 부분은 undefined error 발생 가능.

수정
```
originalRequest.headers = originalRequest.headers ?? {};
originalRequest.headers["Authorization"] = `Bearer ${newAccessToken}`;
```

kimingzero commented on src/utils/jwtUtils.ts:
현재 코드의 개선 필요 포인트
1) jwtDecode 타입이 any라 타입 안전성이 낮음

jwtDecode(token) 결과가 exp를 보장하지 않음 → 타입 지정 필요

2) exp * 1000 < Date.now() + 10000 (10초 버퍼)

이게 좋은 패턴이긴 한데
서버 시차, 모바일 시계 지연 등으로 false-positive가 생길 수 있어.

➡ 일반적으로 1초~3초 정도가 적절하고, 10초는 꽤 큰 편.

(물론 서비스 상황 따라 다름)

3) exp가 number인데 혹시라도 string으로 들어오면 에러 발생 가능성
✅ 개선된 최종 버전 (안전성 + 타입 보완 + 예외 처리 강화)
```
import { jwtDecode } from "jwt-decode";

type JwtPayload = {
  exp?: number;
};

export function isTokenExpired(token: string | null): boolean {
  if (!token) return true;

  try {
    const decoded = jwtDecode<JwtPayload>(token);

    // exp 없으면 만료된 것으로 처리
    if (!decoded.exp || typeof decoded.exp !== "number") {
      return true;
    }

    // 서버 시간 차 등을 고려해 3초 정도 여유 버퍼
    const buffer = 3000;

    const now = Date.now();
    const expireTime = decoded.exp * 1000;

    return expireTime < now + buffer;
  } catch (e) {
    // 토큰 형식이 잘못된 경우 → 무조건 만료 처리
    return true;
  }
}
```

kimhm0828 committed:
[refactor] 코드 리뷰 반영

- 상태 관리 변수 일원화
- token이 null이면 바로 refresh 로직 실행하도록 변경
- headers가 null일 경우 초기화하는 로직 추가
- refresh 로직 함수로 분리, 코드 중복 제거
- refresh 요청만을 위한 axios 클라이언트 생성
- originalRequest의 error가 undefined일 경우에 대한 초기화 로직 추가
- 응답 인터셉터에서 refresh 시도할 시 원본 요청을 깊은 복사해서 보내도록
  수정

kimingzero commented on src/utils/axiosInstance.ts:
필수 수정 2가지 (안 하면 오류 발생 가능성 매우 높음)
① config.headers undefined 에러

현재 코드에서 request 인터셉터에 아래가 있어:
```
if (config.skipAuth) {
  config.headers.Authorization = undefined;
}
```

그리고 아래도:
```
config.headers["Authorization"] = `Bearer ${token}`;
```

❌ 문제: config.headers 자체가 undefined일 수 있음 → 런타임에서 “Cannot read properties of undefined” 터짐.

➡ 이 부분은 반드시 수정해야 함.

✔ 필수 수정 코드
```
config.headers = config.headers ?? {};
```

이 한 줄을 인터셉터 최상단에 넣으면 오류 발생 가능성이 100% 사라짐.

② window.location.href 사용 시 SSR 환경 에러 가능성

지금:
```
window.location.href = "/login";
```

❌ 문제: Next.js App Router 환경에서는
서버에서 실행될 수도 있기 때문에 SSR 환경에서는 window가 없음 → ReferenceError 발생 가능.

➡ 이 부분도 필수로 수정해야 함.

✔ 필수 수정 코드
```
if (typeof window !== "undefined") {
  window.location.href = "/login";
}
```

kimingzero commented on src/utils/jwtUtils.ts:
```diff
@@ -1,13 +1,15 @@
+    if (typeof decoded.exp !== 'number') {
+      return true; // 만료 시간 정보가 없으면 만료된 것으로 간주
+    }
+    return decoded.exp * 1000 < Date.now() + 10000;  // 만료 시간과 현재 시간 비교
```
반드시 수정해야 하는 하나
🔥 (1) 버퍼 시간 10초는 너무 큼 → 실제 서비스에서 false-expired 발생 가능성
```
return decoded.exp * 1000 < Date.now() + 10000;
```

10초면 꽤 큰 값이야.

왜 문제?

사용자가 느리게 요청 보내면
→ 사실 유효한 토큰인데도 만료라고 판단해 /auth/refresh 를 불필요하게 호출하게 됨.

모바일 시간 지연이 있는 경우 더 자주 refresh 발생

refresh 트래픽 증가, 서버 부하 증가

✔ 추천 버퍼값: 2~3초 (업계 표준)
```
const BUFFER_MS = 3000; // 3초 버퍼
return decoded.exp * 1000 < Date.now() + BUFFER_MS;
```

👉 3초면 대부분의 clock drift + 네트워크 지연을 커버하면서
false-positive(잘못된 만료 판정)을 거의 제거함.

kimhm0828 committed:
[refactor] 코드 리뷰 반영 - 버퍼 시간 수정 및 config header undefined 문제 수정

kimingzero reviewed (approved):
승인해둘게용
